### PR TITLE
hiveconfig: allow choosing dns record type for private link

### PR DESCRIPTION
**Pull request by abhinavdahiya**
Currently the private link controller creates an ALIAS record in the PHZ
created for customer's cluster. This record points enables the dns
resolution of customer clsuters k8s API to the VPC endpoint, therefore
allowing hive to transparently accessing the cluster over the private
link.

In some cases the ALIAS record cannot be used. For example, in GovCloud
environments the Route53 private hosted zones do not support ALIAS
record type and therefore we have to use A records.

CNAME records would have been more suitable replacement for ALIAS
records, but
CNAME records cannot be created at the apex of a DNS zone like ALIAS
records. see https://stackoverflow.com/a/20230670
The private link architecture uses a DNS zone like
`api.<clustername>.<clusterdomain>`
so that there is no conflicts with authority on DNS resolution. Since
CNAME
records are not supported on the apex, we use A records pointing the
cluster's
API DNS name to the IP addresses of the VPC endpoint directly. Since
these IP
addresses do not change as they are backed by Elastic Network Interfaces
in the
corresponding subnets of the VPC endpoint, this DNS record should remain
stable.

When in A record mode, the controller extracts the ip addresses of the
vpc endpoint attached elastic networking interfaces and uses those ips
to create A record for the cluster's api dns name in the private hosted
zone.

For simplicity and current desired use-case, the dns record type can be
configured globally for all clusters managed by privatelink controller.

```yaml
spec:
  awsPrivateLink:
    dnsRecordType: (Alias | ARecord)
```

xref: https://issues.redhat.com/browse/HIVE-1572

/assign @dgoodwin

### CODE REVIEW

**Comment by abhinavdahiya** [comment]
/hold
i want to add some developing steps that allow anyone to test with private link.

**Comment by abhinavdahiya** [comment]
/hold cancel

included a section on how to setup dev enviornment for private link network.

File: docs/awsprivatelink.md [hunk 1]
```diff
@@ -223,5 +223,220 @@ expectations of required permissions for these credentials.
...
+  awsPrivateLink:
+    dnsRecordType: (Alias | ARecord)
+```
+
 [aws-private-link-overview]: https://docs.aws.amazon.com/vpc/latest/privatelink/endpoint-services-overview.html
 [control-access-vpc-endpoint]: https://docs.aws.amazon.com/vpc/latest/privatelink/vpc-endpoints-access.html#vpc-endpoints-security-groups
+
+## Developing for Private Link
+
+Since Private link requires some networking setup before Hive controller can create
```
Comment (2uasimojo) [inline docs/awsprivatelink.md:261]:
```suggestion
Private link requires some networking setup before Hive controller can create
```

File: docs/awsprivatelink.md [hunk 2]
```diff
@@ -223,5 +223,220 @@ expectations of required permissions for these credentials.
...
+clusters with Private link routing.
+
+### Setup VPCs in AWS
+
+Create 2 sets of VPCs in the same region in an AWS account. One VPC will
+be used for creating an OCP cluster for Hive, and the another will be used
```
Comment (2uasimojo) [inline docs/awsprivatelink.md:267]:
```suggestion
be used for creating an OCP cluster for Hive, and the other will be used
```

File: docs/awsprivatelink.md [hunk 3]
```diff
@@ -223,5 +223,220 @@ expectations of required permissions for these credentials.
...
+as inventory for VPC endpoints.
+
+1. Create a Hive VPC using cloudformation template.
+
+```
+aws --region us-east-1 cloudformation create-stack \
+  --stack-name hive-vpc \
+  --template-body "file://hack/awsprivatelink/vpc.yaml" \
```
Comment (2uasimojo) [inline docs/awsprivatelink.md]: Is there a sample of this file somewhere?

File: docs/awsprivatelink.md [hunk 4]
```diff
@@ -223,5 +223,220 @@ expectations of required permissions for these credentials.
...
+Now update the Route tables using [AWS documentation][aws-vpc-peering-route-tables]
+
+The cloudformation stack creates one route table for all the public subnets, and one route table
+for each private subnet.
+
+First update the private route tables for Hive VPC. Update the routes so that `10.1.0.0/16`
+prefix routes to the VPC peering connection.
+
+Second update the private route tables for Private link VPC. Update the routes so that `10.0.0.0/16`
+prefix routes to the VPC peering connection.
```
Comment (2uasimojo) [inline docs/awsprivatelink.md:384]: These CIDRs appear backwards from how they were set up above, where the [hive VPC](https://github.com/openshift/hive/pull/1476/files#diff-5fed0fb5d73ed353edbb9348b0aaefcb3ff6930752f3488acc253d8e70f198d4R276) was `10.0.0.0/16` and the [private link VPC](https://github.com/openshift/hive/pull/1476/files#diff-5fed0fb5d73ed353edbb9348b0aaefcb3ff6930752f3488acc253d8e70f198d4R311) was `10.1.0.0/16`.

(It's entirely possible I'm misunderstanding, and this is intentionally criss-crossed.)

File: docs/awsprivatelink.md [hunk 5]
```diff
@@ -223,5 +223,220 @@ expectations of required permissions for these credentials.
...
+
+So we want the workers in the Hive cluster to be able to communicate with the VPC
+endpoints that will be created in the Private link VPC. Secondly, we also want the
+workers to receive traffic from the VPC endpoints.
+
+You can use security groups from peered VPCs to allow flow of traffic using [AWS documentation][aws-vpc-peering-security-groups]
+
+1. Find the security group for workers in Hive Cluster
+
+You the Hive VPC id to list the security groups.
```
Comment (2uasimojo) [inline docs/awsprivatelink.md:396]:
```suggestion
Use the Hive VPC id to list the security groups.
```
?

File: docs/awsprivatelink.md [hunk 6]
```diff
@@ -223,5 +223,220 @@ expectations of required permissions for these credentials.
...
+
+```
+aws --region us-east-1 ec2 describe-security-groups \
+    --filters Name=vpc-id,Values=$PRIVATE_LINK_VPC_ID,Name=group-name,Values=default \
+    --query "SecurityGroups[*].{GroupName:GroupName,GroupId:GroupId}"
+```
+
+3. Allow all incoming traffic from Hive cluster's worker security group to VPC endpoints
+
+Create a incoming security group rule in Private link VPC's default security group that allows ALL traffic
```
Comment (2uasimojo) [inline docs/awsprivatelink.md:414]:
```suggestion
Create an incoming security group rule in Private link VPC's default security group that allows ALL traffic
```

File: docs/awsprivatelink.md [hunk 7]
```diff
@@ -223,5 +223,220 @@ expectations of required permissions for these credentials.
...
+
+```
+aws --region us-east-1 ec2 authorize-security-group-ingress \
+    --group-id $PRIVATE_LINK_VPC_DEFAULT_SG --protocol all --port -1 \
+    --source-group $HIVE_CLUSTER_WORKER_SG
+```
+
+4. Allow all incoming traffic from VPC endpoints to Hive cluster's worker security group
+
+Create a incoming security group rule in Hive cluster's worker security group that allows ALL traffic
```
Comment (2uasimojo) [inline docs/awsprivatelink.md:425]:
```suggestion
Create an incoming security group rule in Hive cluster's worker security group that allows ALL traffic
```

File: docs/enhancements/external-cluster-for-pods.md [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+Hive uses an external Kubernetes cluster for its compute needs like running pods
+for extracting information from the release image, provisioning the cluster using
+openshift-install binary or deprovision of the clusters.
+
+## Background
+
+Currently Hive uses the same Kubernetes cluster where it's running to schedule any
+compute that is required. This adds a requirement that the Kubernetes API used by
+Hive must have compute capabilities in addition to being API backend for its
+resources. New projects like [kcp-dev]() allow running a Kubernetes API that has no
```
Comment (2uasimojo) [inline docs/enhancements/external-cluster-for-pods.md]:
```suggestion
resources. New projects like [kcp](https://github.com/kcp-dev/kcp/) allow running a Kubernetes API that has no
```

File: docs/enhancements/external-cluster-for-pods.md [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+compute capabilities, making them a better target for scaling their use for API backend
+only. Therefore if Hive can separate the use of Kubernetes APIs by function i.e. API
+backend (API cluster) and compute backend (compute cluster) which it allows Hive to
+utilize the new possible backends.
+
+## Goals
+
+1. Hive can be configured to use external Kubernetes cluster for it's compute
```
Comment (2uasimojo) [inline docs/enhancements/external-cluster-for-pods.md]:
```suggestion
1. Hive can be configured to use external Kubernetes cluster for its compute
```

File: docs/enhancements/external-cluster-for-pods.md [hunk 10]
```diff
@@ -0,0 +1,229 @@
+# Using External cluster for running pods
```
Comment (2uasimojo) [inline docs/enhancements/external-cluster-for-pods.md]: This file looks unrelated, like it was added to this PR by accident?

File: hack/awsprivatelink/vpc.cf.yaml [hunk 11]
```diff
@@ -0,0 +1,265 @@
+AWSTemplateFormatVersion: 2010-09-09
```
Comment (2uasimojo) [inline hack/awsprivatelink/vpc.cf.yaml:1]: I'm having to take this file 100% on faith.

File: pkg/controller/awsprivatelink/awsprivatelink_controller.go [hunk 12]
```diff
@@ -884,35 +885,74 @@ func (r *ReconcileAWSPrivateLink) reconcileHostedZone(awsClient *awsClient,
...
+				Action:            aws.String(route53.ChangeActionUpsert),
+				ResourceRecordSet: rSet,
 			}},
 		},
 	})
 	if err != nil {
-		hzLog.WithField("aliasDNSName", endpointDNSName).
-			WithField("aliasHostedZone", endpointDNSHostedZone).
-			WithError(err).Error("error add record to Hosted Zone for VPC Endpoint")
+		hzLog.WithError(err).Error("error adding record to Hosted Zone for VPC Endpoint")
```
Comment (2uasimojo) [inline pkg/controller/awsprivatelink/awsprivatelink_controller.go:905]: Does the returned error in this case contain enough information that we don't need to include, say, the `rSet` as a `%+v`?

File: pkg/controller/awsprivatelink/awsprivatelink_controller.go [hunk 13]
```diff
@@ -884,35 +885,74 @@ func (r *ReconcileAWSPrivateLink) reconcileHostedZone(awsClient *awsClient,
...
+		}
+		if len(res.NetworkInterfaces) == 0 {
+			return nil, errors.New("No network interfaces attached to the vpc endpoint")
+		}
+
+		var ips []string
+		for _, eni := range res.NetworkInterfaces {
+			ips = append(ips, aws.StringValue(eni.PrivateIpAddress))
+		}
+		sort.Strings(ips)
```
Comment (2uasimojo) [inline pkg/controller/awsprivatelink/awsprivatelink_controller.go:937]: Why sorting? So we don't spuriously update when the values differ only by order?

Reason I ask: Sorting IPs stringily isn't a great order, as e.g. `9.0.0.1` will sort after `10.0.0.1`. But if it's just to make ordering consistent, it doesn't matter.

[Later] I guess it's for ease of testing?

**Review by 2uasimojo — COMMENTED** [review]
This seems fine as far as my understanding goes. I'm mostly having to take on faith that it works and does what it says it does. A handful of questions inline, and some typos.

File: docs/awsprivatelink.md (hunk 4 above)
Comment (abhinavdahiya) [reply docs/awsprivatelink.md:384]: we need these VPCs to send traffic to other's range using peering connection, that is why they are adding route for the others network range to themselves.
so we need the Hive VPC route tables to include route for 10.1.0.0/16 private link VPC using the peering connection.
and private link VPC route tables to include route for 10.0.0.0/16 Hive VPC using peering connection.

File: hack/awsprivatelink/vpc.cf.yaml (hunk 11 above)
Comment (abhinavdahiya) [reply hack/awsprivatelink/vpc.cf.yaml:1]: it's a pretty standard VPC setup cloudformation for openshift used by QE

File: pkg/controller/awsprivatelink/awsprivatelink_controller.go (hunk 12 above)
Comment (abhinavdahiya) [reply pkg/controller/awsprivatelink/awsprivatelink_controller.go:905]: the error from ChangeResourceRecordSets should give all the info needed to understand the cause of the problem.
the type of record type is also ready global, and therefore that info isn't required to be included here.

File: pkg/controller/awsprivatelink/awsprivatelink_controller.go (hunk 13 above)
Comment (abhinavdahiya) [reply pkg/controller/awsprivatelink/awsprivatelink_controller.go:937]: yes
a) testing and
b) making sure the list of ips included in the dns record is stable to prevent un necessary changes.

**Review by 2uasimojo — COMMENTED** [review]
/lgtm

**Comment by abhinavdahiya** [comment]
> https://github.com/openshift/hive/compare/f5d52f54ff6984767af326cbbd15d3f1c3595d44..d6f5956a8f59994cc6e80c8acacf83b93fff80bc

had to push to include the vendor update after updating the doc comments. @2uasimojo might need a lgtm again 😇
